R/CDProtocol.R

```diff
@@ -0,0 +1,133 @@
+      } else {
+        params_names <- ls(params_env)
+      }
+      params_optional <- sapply(params_names, function(name) isTRUE(params_env[[name]]$optional))
+      if(length(params_optional) > 0) {
+        params_optional <- sort(params_optional) # get the required params first
+      }
+      params_optional <- c(params_optional, callback = TRUE)
+      # build the pairlist string:
+      text <- paste0(
+        "alist(",
+        paste(names(params_optional), ifelse(params_optional, "NULL", ""), sep = " = ", collapse = ", "),
+        ")"
+      )
+      # return the pairlist:
+      eval(parse(text = text))
+    },
+    list_events = function(domain) {
+      private$.list_objects(domain, "events")
+    },
+    list_types = function(domain) {
+      private$.list_objects(domain, "types")
+    },
+    domains = NULL
+  ),
+  private = list(
+    .protocol = "environment",
+    .list_objects = function(domain, cl) {
+      specs <- get(domain, private$.protocol$domains)[[cl]]
+      if(is.null(specs)) return(list())
+      obj_names <- ls(specs)
+      l <- as.list(paste(domain, obj_names, sep = "."))
+      names(l) <- obj_names
+      l
+    }
+  )
+)
+
+from_json <- function(path) {
+  jsonlite::fromJSON(
+    path,
+    simplifyVector = TRUE,
+    simplifyDataFrame = FALSE,
+    simplifyMatrix = FALSE
+  )
+}
+
+read_local_protocol <- function() {
+  js_protocol <- from_json(local_protocol_file("js"))
+  browser_protocol <- from_json(local_protocol_file("browser"))
+  browser_protocol$domains <- c(browser_protocol$domains, js_protocol$domains)
+  browser_protocol
+}
+
+local_protocol_file <- function(file = c("js", "browser")) {
+  file <- match.arg(file)
+  system.file("protocol", paste0(file, "_protocol.json"), package = "crrri", mustWork = TRUE)
+}
+
+rlist2env <- function(.l) {
+  if (!is.list(.l)) return(.l)
+  list2env(lapply(.l, rlist2env), parent = emptyenv())
+}
+
+add_names_to_protocol <- function(protocol) {
+  protocol$domains <- lapply(protocol$domains, add_names_to_domain)
+  names(protocol$domains) <- sapply(protocol$domains, function(x) x$domain)
+  protocol
+}
+
+add_names_to_domain <- function(domain) {
+  if(!is.null(domain$dependencies)) {
+    domain$dependencies <- unlist(domain$dependencies)
+  }
+  if(!is.null(domain$types)) {
+    domain$types <- lapply(domain$types, add_names_to_type)
+    names(domain$types) <- sapply(domain$types, function(x) x$id)
+  }
+  if(!is.null(domain$commands)) {
+    domain$commands <- lapply(domain$commands, add_names_to_method)
+    names(domain$commands) <- get_names(domain$commands)
+  }
+  if(!is.null(domain$events)) {
+    domain$events <- lapply(domain$events, add_names_to_method)
+    names(domain$events) <- get_names(domain$events)
+  }
+  domain
+}
+
+add_names_to_type <- function(type) {
+  if(!is.null(type$properties)) {
+    names(type$properties) <- get_names(type$properties)
+  }
+  type
+}
+
+add_names_to_method <- function(method) {
```

Comment:
I have the impression that there is some refactoring to do with all those `add_names_to_*` method, and `if(!is.null()) name() <- get_names()`. I'll keep that in mind.

Reply:
I think so. Writing this stuff was painful and I did not finish to factorize all the functions.

Reply:
Ok this is something I can help with. I suggest : 

* leave it as is currently because it works 
* Merged the PR
* Open an issue about refactoring and deal with that later. 

We Could do the same for other improvement we see. 

Otherwise, we will spend like forever in this branch 😄 

Do you agree ?

Reply:
Sounds like a good plan

Reply:
We'll deal with that after merging in #42